### PR TITLE
Tests for ESDT new feature

### PR DESCRIPTION
**Pull request by miiu96**
- Added unit tests for files `esdtLocalBurn.go`, `esdtLocalMint.go`, `esdtRoles.go`
- Added unit tests for file `esdt.go`
- Added semi-integrations tests for the `local mint` built in function and the `local burn` built in function
- Added integration tests for the new functionalities

### CODE REVIEW

**Review by sasurobert — APPROVED** [review]

File: process/smartContract/builtInFunctions/esdtLocalMint.go [hunk 1]
```diff
@@ -60,25 +61,26 @@ func (e *esdtLocalMint) SetNewGasConfig(gasCost *process.GasCost) {
...
-	err = e.rolesHandler.CheckAllowedToExecute(acntSnd, esdtTokenKey, []byte(core.ESDTRoleLocalMint))
+	tokenID := vmInput.Arguments[0]
+	err = e.rolesHandler.CheckAllowedToExecute(acntSnd, tokenID, []byte(core.ESDTRoleLocalMint))
 	if err != nil {
 		return nil, err
 	}
 
 	value := big.NewInt(0).SetBytes(vmInput.Arguments[1])
-	err = addToESDTBalance(vmInput.CallerAddr, acntSnd, esdtTokenKey, big.NewInt(0).Set(value), e.marshalizer, e.pauseHandler)
+	sdtTokenKey := append(e.keyPrefix, tokenID...)
```
Comment (sasurobert) [inline process/smartContract/builtInFunctions/esdtLocalMint.go:82]: esdtTokenKey

**Review by sasurobert — CHANGES_REQUESTED** [review]

File: process/smartContract/builtInFunctions/esdtLocalMint_test.go [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+	}{
+		{
+			name: "NilMarshalizer",
+			argsFunc: func() (c uint64, m marshal.Marshalizer, p process.ESDTPauseHandler, r process.ESDTRoleHandler) {
+				return 0, nil, &mock.PauseHandlerStub{}, &mock.ESDTRoleHandlerStub{}
+			},
+			exError: process.ErrNilMarshalizer,
+		},
+		{
+			name: "NilPouseHandler",
```
Comment (SebastianMarian) [inline process/smartContract/builtInFunctions/esdtLocalMint_test.go:33]: NilPauseHandler instead NilPouseHandler

File: process/smartContract/builtInFunctions/esdtRoles_test.go [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+		VMInput: vmcommon.VMInput{
+			CallValue:  big.NewInt(0),
+			CallerAddr: vm.ESDTSCAddress,
+			Arguments:  [][]byte{[]byte("1"), []byte(core.ESDTRoleLocalMint)},
+		},
+	})
+	require.Equal(t, localErr, err)
+}
+
+func TestEsdtRoles_ProcessBuiltinFunction_UnsetRolesDoesNotExitsShouldWork(t *testing.T) {
```
Comment (SebastianMarian) [inline process/smartContract/builtInFunctions/esdtRoles_test.go:152]: Exist instead Exits

File: process/smartContract/builtInFunctions/esdtRoles_test.go [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+					}
+					return marshalizer.Marshal(roles)
+				},
+			}
+		},
+	}, []byte("ID"), []byte(core.ESDTRoleLocalMint))
+	require.Nil(t, err)
+}
+
+func TestEsdtRolest_CheckAllowedToExecuteRoleNotFind(t *testing.T) {
```
Comment (SebastianMarian) [inline process/smartContract/builtInFunctions/esdtRoles_test.go:286]: Roles instead Rolest

File: vm/systemSmartContracts/esdt_test.go [hunk 5]
```diff
@@ -2267,3 +2268,552 @@ func getAddress() []byte {
...
+	vmInput.Arguments = [][]byte{[]byte("myToken"), []byte("caller"), []byte(core.ESDTRoleLocalBurn)}
+	vmInput.CallerAddr = []byte("caller")
+	vmInput.CallValue = big.NewInt(0)
+	vmInput.GasProvided = 50000000
+
+	retCode := e.Execute(vmInput)
+	require.Equal(t, vmcommon.UserError, retCode)
+}
+
+func TestEsdt_SetSpecialRoleAlreadyExits(t *testing.T) {
```
Comment (SebastianMarian) [inline vm/systemSmartContracts/esdt_test.go:2347]: Exists instead Exits

File: vm/systemSmartContracts/esdt_test.go [hunk 6]
```diff
@@ -2267,3 +2268,552 @@ func getAddress() []byte {
...
+
+	e, _ := NewESDTSmartContract(args)
+	vmInput := getDefaultVmInputForFunc("getAllAddressesAndRoles", [][]byte{})
+	vmInput.Arguments = nil
+
+	retCode := e.Execute(vmInput)
+	require.Equal(t, vmcommon.UserError, retCode)
+}
+
+func TestEsdt_GetAllAddressesAndRoleCallWithValueShouldErr(t *testing.T) {
```
Comment (SebastianMarian) [inline vm/systemSmartContracts/esdt_test.go:2766]: AndRoles

File: vm/systemSmartContracts/esdt_test.go [hunk 7]
```diff
@@ -2267,3 +2268,552 @@ func getAddress() []byte {
...
+	e, _ := NewESDTSmartContract(args)
+	vmInput := getDefaultVmInputForFunc("getAllAddressesAndRoles", [][]byte{})
+	vmInput.Arguments = [][]byte{[]byte("arg")}
+	vmInput.CallValue = big.NewInt(0)
+
+	retCode := e.Execute(vmInput)
+	require.Equal(t, vmcommon.UserError, retCode)
+}
+
+func TestEsdt_GetAllAddressesAndRoleCallGetExistingTokenErr(t *testing.T) {
```
Comment (SebastianMarian) [inline vm/systemSmartContracts/esdt_test.go:2782]: AndRoles

File: vm/systemSmartContracts/esdt_test.go [hunk 8]
```diff
@@ -2267,3 +2268,552 @@ func getAddress() []byte {
...
+	vmInput.Arguments = [][]byte{[]byte("myToken"), []byte("caller"), []byte(core.ESDTRoleLocalBurn)}
+	vmInput.CallerAddr = []byte("caller")
+	vmInput.CallValue = big.NewInt(0)
+	vmInput.GasProvided = 50000000
+
+	retCode := e.Execute(vmInput)
+	require.Equal(t, vmcommon.UserError, retCode)
+}
+
+func TestEsdt_UnsetSpecialRoleCannotRemoveRoleNotExitsShouldErr(t *testing.T) {
```
Comment (SebastianMarian) [inline vm/systemSmartContracts/esdt_test.go:2584]: Exists instead Exits

File: process/smartContract/builtInFunctions/esdtLocalMint.go (hunk 1 above)
Comment (miiu96) [reply process/smartContract/builtInFunctions/esdtLocalMint.go:82]: done.

File: process/smartContract/builtInFunctions/esdtLocalBurn_test.go [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+	}{
+		{
+			name: "NilMarshalizer",
+			argsFunc: func() (c uint64, m marshal.Marshalizer, p process.ESDTPauseHandler, r process.ESDTRoleHandler) {
+				return 0, nil, &mock.PauseHandlerStub{}, &mock.ESDTRoleHandlerStub{}
+			},
+			exError: process.ErrNilMarshalizer,
+		},
+		{
+			name: "NilPouseHandler",
```
[the remark this answers is not in the document]
Comment (miiu96) [reply process/smartContract/builtInFunctions/esdtLocalBurn_test.go:33]: done.

File: process/smartContract/builtInFunctions/esdtLocalMint_test.go (hunk 2 above)
Comment (miiu96) [reply process/smartContract/builtInFunctions/esdtLocalMint_test.go:33]: done.

File: process/smartContract/builtInFunctions/esdtRoles_test.go (hunk 3 above)
Comment (miiu96) [reply process/smartContract/builtInFunctions/esdtRoles_test.go:152]: done.

File: process/smartContract/builtInFunctions/esdtRoles_test.go (hunk 4 above)
Comment (miiu96) [reply process/smartContract/builtInFunctions/esdtRoles_test.go:286]: done.

File: vm/systemSmartContracts/esdt_test.go (hunk 6 above)
Comment (miiu96) [reply vm/systemSmartContracts/esdt_test.go:2766]: done.

File: vm/systemSmartContracts/esdt_test.go (hunk 5 above)
Comment (miiu96) [reply vm/systemSmartContracts/esdt_test.go:2347]: done.

File: vm/systemSmartContracts/esdt_test.go (hunk 7 above)
Comment (miiu96) [reply vm/systemSmartContracts/esdt_test.go:2782]: done/

File: vm/systemSmartContracts/esdt_test.go (hunk 8 above)
Comment (miiu96) [reply vm/systemSmartContracts/esdt_test.go:2584]: done.

**Review by sasurobert — DISMISSED** [review]

**Review by SebastianMarian — APPROVED** [review]

**Review by mateigraura — APPROVED** [review]
System test passed

**Review by sasurobert — APPROVED** [review]

**Pull request merged**